src/Nethermind/Nethermind.Consensus/Processing/ParallelProcessing/MultiVersionMemory.cs

```diff
@@ -0,0 +1,340 @@
+        {
+            DataDictionary<TLocation, Value> txData = _data[txIndex];
+            txData.Lock.EnterWriteLock();
+            foreach (var location in previousLocations)
+            {
+                txData.Dictionary[location] = Value.Estimate;
+            }
+            txData.Lock.ExitWriteLock();
+        }
+    }
+
+    /// <summary>
+    /// Tries to read the location from memory.
+    /// </summary>
+    /// <param name="location">Location to read</param>
+    /// <param name="txIndex">Index of transaction that is reading</param>
+    /// <param name="version">out version that was read - information about transaction that written to this location</param>
+    /// <param name="value">Value that was read</param>
+    /// <returns>
+    /// <see cref="Status.Ok"/> when read succeeds.
+    /// <see cref="Status.NotFound"/> when value for location is not in memory and needs to be read from database.
+    /// <see cref="Status.ReadError"/> when we know previous transaction needs to be re-executed first to correctly read the location.
+    /// </returns>
+    public Status TryRead(TLocation location, ushort txIndex, out Version version, out byte[]? value)
+    {
+        long id = parallelTrace.ReserveId();
+        ushort prevTx = txIndex;
+        prevTx--; // start from previous transaction and go back through all the previous transactions
+        while (prevTx != ushort.MaxValue) // hack for ushort underflow, so going < 0
+        {
+            DataDictionary<TLocation, Value> prevTransactionData = _data[prevTx];
+            prevTransactionData.Lock.EnterReadLock();
+            // if we find the location written by previous transaction
+            if (prevTransactionData.Dictionary.TryGetValue(location, out Value v))
+            {
+                prevTransactionData.Lock.ExitReadLock();
+                version = new Version(prevTx, v.Incarnation); // return version info
+
+                if (v.IsEstimate)
+                {
+                    // if estimate (prevTx needs re-execution) return ReadError.
+                    value = null;
+                    if (typeof(TLogger) == typeof(IsTracing)) parallelTrace.Add(id, $"Tx {txIndex} TryRead at location {location} returned {Status.ReadError} with blocking {version}.");
+                    return Status.ReadError;
+                }
+                else
+                {
+                    // else we can return the value
+                    value = v.Bytes;
+                    if (typeof(TLogger) == typeof(IsTracing)) parallelTrace.Add(id, $"Tx {txIndex} TryRead at location {location} returned {Status.Ok} with value {value.ToHexString()} from {version}.");
+                    return Status.Ok;
+                }
+            }
+            else
+            {
+                prevTransactionData.Lock.ExitReadLock();
+            }
+
+            // if not we are going backwards until we find it or checked all transactions
+            prevTx--;
```

Comment:
If a value was set in transaction number 1 with an arbitrary incarnation (not important), and we retrieve it from 60, do we need to spin through all the transaction dictionaries to get the value?

Reply:
No if `prevTransactionData.Dictionary.TryGetValue` is true we have a `return` statement. So we go until we find one, but need precedence of highest index smaller than current transaction.

This definitively could be optimized somehow.